examples/catalan.masm

```diff
@@ -0,0 +1,160 @@
+# Computes the nth Catalan number using dynamic programming.
+# (See https://www.geeksforgeeks.org/program-nth-catalan-number/)
+#
+# Definition:
+#     C_0 = 1
+#     C_(n+1) = C_0 * C_n + C_1 * C_(n-1) + ... + C_n * C_0
+#             =  2 * C_0 * C_n + 2 * C_1 * C_(n-1) + ... + 2 * C_(n/2) * C_(n/2+1) (if n+1 is even)
+#             or 2 * C_0 * C_n + 2 * C_1 * C_(n-1) + ... + C_(n/2)^2               (if n+1 is odd)
+#
+# Input: The stack should contain a single element n.
+# Output: A stack with a single element which is the nth Catalan number C_n.
+#
+# C_i for 0 <= i <= n will be stored in memory at address i.
+#
+# example:
+#
+# input:
+# [9]
+#
+# output:
+# [4862]
+
+proc.dup1dup1
+    dup.1
+    dup.1
+end
+
+# Check if the counter is less than the maximum
+proc.check_continue_gt
+    # [i, max]
+    exec.dup1dup1
+    gt
+    # [ i < max , i, max ]
+end
+
+# Fetch precomputed C_j and C_(i-j), and multiply them
+proc.fetch_and_multiply
+    # stack : [ i, j, ... ], mem : [ C_0, C_1, ..., C_i ]
+    # Fetch C_(i-j)
+    dup.1
+    sub
+    mem_load
+    # Fetch C_j
+    swap
+    mem_load
+    # Multiply
+    mul
+    # [ C_j * C_(i-j), ... ]
+end
+
+# Calculate C_i.
+# Assumes that all C_j for j < i have already been calculated and stored in memory
+proc.catalan_i
+    # stack : [ i, ... ], mem : [ C_0, C_1, ..., C_(i-1) ]
+    dup
+    push.0
+    eq
+    if.true
+        # Base case: C_0 = 1
+        # [ 0, ... ]
+        push.1
+    else
+        # Induction case: C_i = C_(k+1)
+        # [ i, ... ]
+        dup
+        sub.1
+        # [ k, i, ... ]
+        # Initialize running total
+        push.0
+        # [ running_total, k, i, ... ]
+        # Initialize maximum for counter (k/2)
+        dup.1
+        # Need to use u32 division to truncate properly
+        # Safe to use unchecked, since k < i <= n+1, and n+1 has been asserted as u32
+        u32unchecked_div.2
+        # [ k/2, running_total, k, i, ... ]
+        # Initialize counter
+        push.0
+        # [ j, k/2, running_total, k, i, ... ]
+        # Calculate sum until penultimate summand
+        exec.check_continue_gt
```

Comment:
Something to consider here is that `gt` incurs a [high clock cycle overhead](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/field_operations.html#comparison-operations) (18 cycles) compared to `eq` (1 cycle) /`neq` (2 cycles).  You may be able to refactor this check to use `neq`.  I believe by design j starts off as 0 which is lower than `k/2` and hence you can change your continue condition to check for inequality.  Once j == 2 then you know that !(j < k/2).   

```
# Check if the counter is less than the maximum
proc.check_continue_gt
    # [i, max]
    exec.dup1dup1
    gt
    # [ i < max , i, max ]
end
```

would be become

```
# Check if the counter is less than the maximum
proc.check_continue
    # [i, max]
    exec.dup1dup1
    neq
    # [ i < max , i, max ]
end
```

I implemented a similar pattern of using `gt` checks on my first miden script and was advised the same so I am just passing on the wisdom.

Reply:
It's a good catch either way. :-)

I've refactored both loop conditions to use `neq` which simplifies the code as a whole, at the cost of two additional instructions (`add.1` and `sub.1`) in the main function.